src/ParseFasm.cpp

```diff
@@ -0,0 +1,574 @@
+                        }
+                        s.put(c);
+                }
+        } else {
+                s << str.data;
+        }
+        return s;
+}
+
+/// Wraps a string in another header, used to aggregate data.
+std::string withHeader(char tag, std::string data) {
+        std::ostringstream header;
+        header << tag << Num(data.size());
+        return header.str() + data;
+}
+
+/// Counts characters that don't match the given character.
+/// Used to count digits skipping '_'.
+int count_without(std::string::iterator start,
+                  std::string::iterator end,
+                  char c) {
+        int count = 0;
+        auto it = start;
+        while (it != end) {
+                if (*it != c) {
+                        count++;
+                }
+                it++;
+        }
+        return count;
+}
+
+/// Calculates the number of leading pad bits are needed
+/// so that the rightmost bit will be the LSB of a Num.
+/// e.g. This would be 31 for 33'b0.
+int lead_bits(int bits) {
+        return (Num::kWidth - (bits % Num::kWidth)) % Num::kWidth;
+}
+
+// clang-format off
+/// Decode a hex digit.
+int from_hex_digit(char c) {
+    int result =
+        c >= '0' && c <= '9' ? c - '0' :
+        c >= 'a' && c <= 'f' ? c - 'a' :
+        c >= 'A' && c <= 'F' ? c - 'A' : -1;
+    assert(result >= 0 && result < 16);
+    return result;
+}
+// clang-format on
+
+/// Makes tags easy to extract for documentation and code generation.
+/// Use at most once per line to allow simple grepping.
+#define TAG(c, long_name) (c)
+
+/// Raised on parse errors
+struct ParseException {
+        size_t line;          ///< Line number of error.
+        size_t position;      ///< Position in that line.
+        std::string message;  ///< A descriptive message.
+};
+
+/// FasmParserBaseVisitor is a visitor for the parse tree
+/// generated by the ANTLR parser.
+/// It will encode the tree a line at a time and stream out of
+/// the given std::ostream
+class FasmParserBaseVisitor : public FasmParserVisitor {
+       public:
+        static constexpr size_t kHeaderSize = 5;
+
+        /// The constructor requires a std::ostream to stream encoded lines.
+        /// This is to avoid storing an entire copy of the parse tree in a
+        /// different form.
+        FasmParserBaseVisitor(std::ostream& out) : out(out) {}
+
+        /// Stream out FASM lines.
+        virtual Any visitFasmFile(
+            FasmParser::FasmFileContext* context) override {
+                for (auto& line : context->fasmLine()) {
+                        std::string str = visit(line);
+                        if (!str.empty()) {
+                                out << str;
+                                if (hex_mode)
+                                        out << std::endl;
+                        }
+                }
+                return {};
+        }
+
+/// Helper macro to convert a rule context into a string
```

Comment:
A macro definition inside the middle of a class definition is *super* weird and looks like an error.

Reply:
I like to minimize the scope; I usually mark the end of the scope with `#undef X`. But it doesn't matter much here; I can put it at the top if needed.

Reply:
Done. Moved above the class, and added `#undef` below the class.